doc/src/devdocs/llvm-passes.md

```diff
@@ -0,0 +1,159 @@
+
+### CPUFeatures
+
+* Filename: `llvm-cpufeatures.cpp`
+* Class Name: `CPUFeaturesPass`
+* Opt Name: `module(CPUFeatures)`
+
+This pass lowers the `julia.cpu.have_fma.(f32|f64)` intrinsic to either true or false, depending on the target architecture and target features present on the function. This intrinsic is often used to determine if using algorithms dependent on fast [fused multiply-add](https://en.wikipedia.org/wiki/Multiply%E2%80%93accumulate_operation#Fused_multiply%E2%80%93add) operations is better than using standard algorithms not dependent on such instructions.
+
+### DemoteFloat16
+
+* Filename: `llvm-demote-float16.cpp`
+* ClassName: `DemoteFloat16Pass`
+* Opt Name `function(DemoteFloat16)`
+
+This pass replaces [float16](https://en.wikipedia.org/wiki/Half-precision_floating-point_format) operations with float32 operations on architectures that do not natively support float16 operations. This is done by inserting `fpext` and `fptrunc` instructions around any float16 operation. On architectures that do support native float16 operations, this pass is a no-op.
+
+### LateGCLowering
+
+* Filename: `llvm-late-gc-lowering.cpp`
+* Class Name: `LateLowerGCPass`
+* Opt Name: `function(LateLowerGCFrame)`
+
+This pass performs most of the GC rooting work required to track pointers between GC safepoints. It also lowers several intrinsics to their corresponding instruction translation, and is permitted to violate the non-integral invariants previously established (`pointer_from_objref` is lowered to a `ptrtoint` instruction here). This pass typically occupies the most time out of all the custom Julia passes, due to its dataflow algorithm to minimize the number of objects live at any safepoint.
+
+### FinalGCLowering
+
+* Filename: `llvm-final-gc-lowering.cpp`
+* Class Name: `FinalLowerGCPass`
+* Opt Name: `module(FinalLowerGC)`
+
+This pass lowers a few last intrinsics to their final form targeting functions in the `libjulia` library. Separating this from `LateGCLowering` enables other backends (GPU compilation) to supply their own custom lowerings for these intrinsics, enabling the Julia pipeline to be used on those backends as well.
+
+### LowerHandlers
+
+* Filename: `llvm-lower-handlers.cpp`
+* Class Name: `LowerExcHandlersPass`
+* Opt Name: `function(LowerExcHandlers)`
+
+This pass lowers exception handling intrinsics into calls to runtime functions that are actually called when handling exceptions.
+
+### RemoveNI
+
+* Filename: `llvm-remove-ni.cpp`
+* Class Name: `RemoveNIPass`
+* Opt Name: `module(RemoveNI)`
+
+This pass removes the non-integral address spaces from the module's datalayout string. This enables the backend to lower Julia's custom address spaces directly to machine code, without a costly rewrite of every pointer operation to address space 0.
```

Comment:
I assume you meant this in the reverse? "This enables buggy backends to still lower Julia's custom address spaces to machine code, albeit with a costly rewrite of every pointer operation to address space 0."

Reply:
Oh, sorry, I was mixing this up this pass with RemoveAddrspaces